llmfoundry/callbacks/async_eval_callback.py

```diff
@@ -0,0 +1,208 @@
+            use for the eval run. If not provided, the same cluster as the current run and a
+            single GPU node will be used.
+    """
+
+    def __init__(
+        self,
+        interval: Union[str, int, Time],
+        compute: Optional[Union[ComputeConfig, Dict[str, Any]]] = None,
+    ):
+        self.check_interval = create_interval_scheduler(interval)
+        self.compute = compute
+        self.count = 0
+        self.last_launch: Optional[Time] = None
+
+        # Run these during init to fail fast in any of the error cases
+        self.current_run = self._get_current_run()
+        self.get_eval_parameters(
+            self.current_run.submitted_config.parameters or {},
+            self.current_run.name,
+        )
+        log.info(
+            f'Initialized AsyncEval callback. Will generate runs at interval {interval}'
+        )
+
+    def run_event(self, event: Event, state: State, logger: Logger) -> None:
+        del logger
+        if all([
+                state.get_elapsed_duration() is not None,
+                self.check_interval(state, event),
+                self.last_launch != state.timestamp.batch,
+                dist.get_global_rank() == 0
+        ]):
+            self.launch_run()
+
+            self.last_launch = state.timestamp.batch
+            self.count += 1
+
+    def _get_current_run(self) -> Run:
+        if os.environ.get(MOSAICML_PLATFORM_ENV_VAR,
+                          'false').lower() == 'false':
+            raise RuntimeError(
+                'AsyncEval callback is only supported when running on the MosaicML platform'
+            )
+
+        run_name = os.environ.get(RUN_NAME_ENV_VAR, None)
+        if not run_name:
+            raise RuntimeError(
+                'RUN_NAME environment variable must be set to use the AsyncEval callback'
+            )
+
+        # Allows the MapiException to be raised if the run doesn't exist
+        return get_run(run_name, include_details=True)
+
+    def get_eval_parameters(
+        self,
+        parameters: Dict[str, Any],
+        run_name: str,
+    ) -> Dict[str, Any]:
+        looking_for = REQUIRED_PARAMS_FOR_EVAL.copy()
+
+        # Go through all parameters and pull out the ones needed for eval
+        subset_keys = {}
+        for key in parameters:
+            if key in OPTIONAL_PARAMS_FOR_EVAL:
+                subset_keys[key] = parameters[key]
+            elif key in REQUIRED_PARAMS_FOR_EVAL:
+                subset_keys[key] = parameters[key]
+                looking_for.remove(key)
+
+        if looking_for:
+            raise Exception(
+                f'Missing the following required parameters for async eval: {looking_for}'
+            )
+
+        # Convert the save_folder to a load_path
+        subset_keys['load_path'] = get_load_path(
+            subset_keys.pop('save_folder'),
+            parameters.get('save_latest_filename', None))
+
+        # Update the loggers to use the training run name
```

Comment:
Should it be logged as separate runs or the same run? Debating, it could be confusing because the eval runs could sit in pending

Reply:
Yeah, I think it should be the same run. but then, as we were discussing, we need a way to propagate through what "step" to log the eval for.

Reply:
ok, thats fair. one separate eval run is way better I think. Although as long as you use a wandb group for your run, they would all show up in the group and maybe its not so bad.